rust/src/buffer.rs

```diff
@@ -196,23 +187,43 @@ impl MutableBuffer {
         }
     }
 
-    /// Adjust the capacity of this buffer to be at least `new_capacity`.
+    /// Ensures that this buffer has at least `capacity` slots in this buffer. This will
+    /// also ensure the new capacity will be a multiple of 64 bytes.
+    ///
+    /// Returns the new capacity for this buffer.
+    pub fn reserve(&mut self, capacity: usize) -> Result<usize> {
+        if capacity > self.capacity {
+            let new_capacity = bit_util::round_upto_multiple_of_64(capacity as i64);
+            let new_capacity = cmp::max(new_capacity, self.capacity as i64 * 2) as usize;
+            let new_data = memory::reallocate(self.capacity, new_capacity, self.data)?;
+            self.data = new_data as *mut u8;
+            self.capacity = new_capacity;
+        }
+        Ok(self.capacity)
+    }
+
+    /// Resizes the buffer so that the `len` will equal to the `new_len`.
     ///
-    /// If the `new_capacity` is less than the current capacity, nothing is done and `Ok`
-    /// will be returned. Otherwise, the new capacity value will be chosen between the
-    /// larger one of the incoming `new_capacity` (after rounding up to the nearest 64)
-    /// and the doubled value of the existing capacity, the capacity added is returned.
-    pub fn resize(&mut self, new_capacity: usize) -> Result<usize> {
-        if new_capacity <= self.capacity {
-            return Ok(0);
+    /// If `new_len` is greater than `len`, the buffer's length is simply adjusted to be
+    /// the former, optionally extending the capacity. The data between `len` and
+    /// `new_len` will remain unchanged.
+    ///
+    /// If `new_len` is less than `len`, the buffer will be truncated.
+    pub fn resize(&mut self, new_len: usize) -> Result<()> {
+        if new_len > self.len {
```

Comment:
I think this check is redundant as the call to `reserve` will check `new_len` against `self.capacity`

Reply:
I marked the wrong line, I mean this check:
`if new_len > self.capacity {`
on line 214

Reply:
Good point. Removed.